File: frontend/src/services/kpi.ts
import api from "./api";

export interface KPIValue {
  id: number;
  indicator_id: number;
  year: number;
  month: number;
  value_raw: string | null;
  value_numeric: number | null;
}

export interface KPIYearlyTarget {
  id: number;
  indicator_id: number;
  year: number;
  frequency: string | null;
  target_raw: string | null;
  target_numeric: number | null;
  target_numeric_max: number | null;
  operator: "GTE" | "LTE" | "BETWEEN" | "EQ" | null;
}

export interface KPIIndicator {
  id: number;
  processus_id: number;
  name: string;
  periodicity: 'MONTHLY' | 'QUARTERLY' | 'SEMESTERLY';
  yearly_targets: KPIYearlyTarget[];
  values: KPIValue[];
}

export interface KPIProcessus {
  id: number;
  name: string;
  department_id: number | null;
  editor_role_names: string[];
  editor_user_ids: number[];
  indicators: KPIIndicator[];
}

export const kpiService = {
  async uploadPreview(file: File): Promise<string[]> {
    const formData = new FormData();
    formData.append("file", file);
    
    const response = await api.post("/kpi/upload-preview", formData, {
      headers: { "Content-Type": "multipart/form-data" },
    });
    return response.data.sheet_names;
  },

  async uploadParse(
    file: File, 
    sheetName: string, 
    year: number, 
    monthName: string, 
    monthIndex: number
  ): Promise<{ message: string, imported_count: number, updated_count: number }> {
    const formData = new FormData();
    formData.append("file", file);
    formData.append("sheet_name", sheetName);
    formData.append("year", year.toString());
    formData.append("month_name", monthName);
    formData.append("month_index", monthIndex.toString());

    const response = await api.post("/kpi/upload-parse", formData, {
      headers: { "Content-Type": "multipart/form-data" },
    });
    return response.data;
  },

  async getDashboardData(year: number): Promise<KPIProcessus[]> {
    const response = await api.get(`/kpi/dashboard/${year}`);
    return response.data;
  },

  async createProcessus(name: string, department_id?: number | null): Promise<KPIProcessus> {
    const response = await api.post("/kpi/processus", { name, department_id });
    return response.data;
  },

  async createIndicator(name: string, processus_id: number, periodicity: string = 'MONTHLY'): Promise<KPIIndicator> {
    const response = await api.post("/kpi/indicators", { name, processus_id, periodicity });
    return response.data;
  },

  async configureEditors(processus_id: number, editor_role_names: string[], editor_user_ids: number[]): Promise<KPIProcessus> {
    const response = await api.put(`/kpi/processus/${processus_id}/editors`, {
      editor_role_names,
      editor_user_ids
    });
    return response.data;
  },

  async updateKpiValue(indicator_id: number, year: number, month: number, value_raw: string, value_numeric: number | null): Promise<KPIValue> {
    const response = await api.post("/kpi/values", { indicator_id, year, month, value_raw, value_numeric });
    return response.data;
  }
};

File: backend/app/modules/quality/routes/kpi.py
from fastapi import APIRouter, Depends, UploadFile, File, Form, HTTPException
from sqlalchemy.orm import Session
from typing import List

from app.core.database.session import get_db
from app.core.security.auth import get_current_user
from app.modules.users.models.user import User

from app.modules.quality.schemas.kpi import KPIImportPreviewResponse, KPIImportResponse, KPIProcessusResponse, KPIProcessusCreate, KPIIndicatorCreate, KPIIndicatorResponse, KPIValueCreate, KPIValueResponse
from app.modules.quality.services.kpi import get_excel_preview, parse_and_import_kpi, get_kpi_dashboard_data
from app.modules.quality.models.kpi import KPIProcessus, KPIIndicator, KPIValue

router = APIRouter(prefix="/kpi", tags=["Quality KPI"])

@router.post("/upload-preview", response_model=KPIImportPreviewResponse)
def api_kpi_upload_preview(
    file: UploadFile = File(...),
    current_user: User = Depends(get_current_user)
):
    if not any(r.name in ["Admin", "Qualité", "Qualite"] for r in current_user.roles):
        raise HTTPException(status_code=403, detail="Permission refusée")
        
    contents = file.file.read()
    sheet_names = get_excel_preview(contents)
    return {"sheet_names": sheet_names}

@router.post("/upload-parse", response_model=KPIImportResponse)
def api_kpi_upload_parse(
    file: UploadFile = File(...),
    sheet_name: str = Form(...),
    year: int = Form(...),
    month_name: str = Form(...),
    month_index: int = Form(...),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    if not any(r.name in ["Admin", "Qualité"] for r in current_user.roles):
        raise HTTPException(status_code=403, detail="Permission refusée")
        
    contents = file.file.read()
    imported, updated = parse_and_import_kpi(db, contents, sheet_name, year, month_name, month_index)
    
    return {
        "message": f"Extraction réussie. {imported} KPI importés, {updated} mis à jour.",
        "imported_count": imported,
        "updated_count": updated
    }

@router.get("/dashboard/{year}", response_model=List[KPIProcessusResponse])
def api_get_kpi_dashboard(
    year: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    data = get_kpi_dashboard_data(db, year)
    return data

@router.post("/processus", response_model=KPIProcessusResponse)
def create_processus(
    data: KPIProcessusCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    if not any(r.name in ["Admin", "Qualité", "Qualite"] for r in current_user.roles):
        raise HTTPException(status_code=403, detail="Seule la Qualité peut créer un processus.")
    proc = KPIProcessus(**data.model_dump())
    db.add(proc)
    db.commit()
    db.refresh(proc)
    return proc

@router.post("/indicators", response_model=KPIIndicatorResponse)
def create_indicator(
    data: KPIIndicatorCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    if not any(r.name in ["Admin", "Qualité", "Qualite"] for r in current_user.roles):
        raise HTTPException(status_code=403, detail="Seule la Qualité peut créer un indicateur.")
    ind = KPIIndicator(**data.model_dump())
    db.add(ind)
    db.commit()
    db.refresh(ind)
    return ind

@router.post("/values", response_model=KPIValueResponse)
def update_kpi_value(
    data: KPIValueCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db)
):
    is_qualite = any(r.name in ["Admin", "Qualité", "Qualite"] for r in current_user.roles)
    is_pilote = any(r.name in ["Pilote", "Copilote"] for r in current_user.roles)
    
    if not is_qualite and not is_pilote:
        raise HTTPException(status_code=403, detail="Permission refusée.")
        
    indicator = db.query(KPIIndicator).filter(KPIIndicator.id == data.indicator_id).first()
    if not indicator:
        raise HTTPException(status_code=404, detail="Indicateur non trouvé.")
        
    if is_pilote and not is_qualite:
        if indicator.processus.department_id != current_user.department_id:
            raise HTTPException(status_code=403, detail="Vous ne pouvez renseigner que les KPIs de votre direction.")
            
    val = db.query(KPIValue).filter(
        KPIValue.indicator_id == data.indicator_id,
        KPIValue.year == data.year,
        KPIValue.month == data.month
    ).first()
    
    if val:
        val.value_raw = data.value_raw
        val.value_numeric = data.value_numeric
    else:
        val = KPIValue(**data.model_dump())
        db.add(val)
        
    db.commit()
    db.refresh(val)
    return val

